buildSrc/reaper/src/main/java/org/elasticsearch/gradle/reaper/Reaper.java

```diff
@@ -0,0 +1,125 @@
+ * executed with {@link Runtime#getRuntime()#exec}.
+ *
+ * The main method will wait indefinitely on the parent process (Gradle) by
+ * reading from stdin. When Gradle shuts down, whether normally or abruptly, the
+ * pipe will be broken and read will return.
+ *
+ * The reaper will then iterate over the files in the configured directory,
+ * and execute the given commands. If any commands fail, a failure message is
+ * written to stderr. Otherwise, the input file will be deleted. If no inputs
+ * produced errors, the entire input directory is deleted upon completion of reaping.
+ */
+public class Reaper implements Closeable {
+
+    Path inputDir;
+    boolean failed;
+
+    private Reaper(Path inputDir) {
+        this.inputDir = inputDir;
+        this.failed = false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length != 1) {
+            System.err.println("Expected one argument.\nUsage: java -cp reaper.jar Reaper <DIR_OF_REAPING_COMMANDS>");
+            System.exit(1);
+        }
+        Path inputDir = Paths.get(args[0]);
+
+        try (Reaper reaper = new Reaper(inputDir)){
+            System.in.read();
+            reaper.reap();
+        }
+    }
+
+    private void reap() {
+        try {
+            final List<Path> inputFiles = Files.list(inputDir)
+                .filter(p -> p.getFileName().toString().endsWith(".log") == false).collect(Collectors.toList());
+
+            for (Path inputFile : inputFiles) {
+                System.out.println("Process file: " + inputFile);
+                for (String line : Files.readAllLines(inputFile)) {
+                    System.out.println("Running command: " + line);
+                    String[] command = line.split(" ");
+                    Process process = Runtime.getRuntime().exec(command);
+                    int ret = process.waitFor();
+
+                    System.out.println("Stdout: " + readStream(process.getInputStream()));
+                    System.out.println("Stderr: " + readStream(process.getErrorStream()));
+                    if (ret != 0) {
+                        logFailure("Command [" + line + "] failed with exit code " + ret, null);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            logFailure("Failed to reap inputs", e);
+        }
+    }
+
+    private void logFailure(String message, Exception e) {
+        System.err.println(message);
+        e.printStackTrace(System.err);
+        failed = true;
+    }
+
+    private static String readStream(InputStream s) throws IOException {
+        ByteArrayOutputStream o = new ByteArrayOutputStream();
+        s.transferTo(o);
+        return o.toString(StandardCharsets.UTF_8);
+    }
+
+    @Override
+    public void close() {
+        if (failed == false) {
+            try {
+                for (Path toDelete : Files.walk(inputDir).sorted(Comparator.reverseOrder()).collect(Collectors.toList())) {
```

Comment:
That won't work, since we need the root directory to be deleted last, otherwise it will fail. Thus we need the reverse sort to guarantee the root comes last.

Reply:
Can't we do `.forEach()` after the sort?

Reply:
Nope, because Files.delete throws IOException, and there is no functional interface that throws exceptions (we have them in ES, but not in our build code). So we would have to add a try/catch within the lambda, which I thought would make this more difficult to understand, and we would still have to wrap it in an unchecked exception, which would just pop back out to the catch here.